designs/2024-baseline-support/README.md

```diff
@@ -0,0 +1,319 @@
+how to generate the baseline and save the results to `/home/user/project/mycache`.
+
+``` bash
+eslint --baseline --baseline-location /home/user/project/mycache
+```
+
+To introduce the above-mentioned options, we will need to: 
+
+ * add the new options in `default-cli-options.js`.
+ * adjust the config for optionator.
+ * add the new options as comments and arguments for eslint.
+ * update documentation to explain the newly introduced options.
+
+On top of that, we will need to adjust `cli.js` to check if `--baseline` was provided and set to true, right after the fixes are done and the warnings are filtered out, to avoid counting errors that are about to be fixed. A new method `generateBaseline` will be introduced, which must be called only and only if `--baseline` was provided and set to true. Please refer to "Implementation Details" for more details on this.
+
+### Matching against the baseline
+
+The suggested solution always compares against the baseline, given that the baseline file exists. By default the baseline file is picked up from `.eslint-baseline.json`, unless the option `--baseline-location` is provided. This makes it easier for existing and new projects to adopt the baseline without the need to adjust scripts in `package.json` and CI/CD workflows. As mentioned above, `--baseline-location` is a string argument specifying where the baseline was previously saved.
+
+This will go through each result item and message, and check each rule giving an error (`severity == 2`) against the baseline. By design, we do not take warnings into consideration (regardless of whether quite mode is enabled or not), since warnings do not cause eslint to exit with an error code and they already serve a different purpose. If the file and rule are part of the baseline, means that we can remove and ignore the result message. 
+
+To implement this, we will need to adjust further `cli.js` and check if the baseline file exists - taking `--baseline-location` into consideration if exists, otherwise fallback to `.eslint-baseline.json`. This needs to take place right after the baseline is generated so that we take the baseline into consideration, if it was just generated. It also needs to take place before the error counting, so that the remaining errors are counted correctly.  
+
+We can also keep track of which errors from baseline were not matched, that is useful for the next section.
+
+### Maintaining a lean baseline
+
+When using the baseline, there is a chance that an error is resolved but the baseline file is not updated. This might allow new errors to creep in without noticing. To ensure that the baseline is always up to date, eslint can exit with an error code when there are ignored errors that do not occur anymore. To fix this, the developer can regenerate the baseline file.
```

Comment:
By _resolved_ you mean _reported_? If an error that is not filtered out by the baseline is reported, I assume that ESLint will display the error message and exit with an error code as usual. It is not clear to me why that would go unnoticed.

Reply:
@fasttime By _resolved_ I am referring to a newly addressed violation. I have added a scenario as an example. 

The purpose of this section, is to encourage developers re-generating the baseline after addressing one or more violations.

Reply:
Okay, thanks. Actually we don't know which errors are resolved. We can only compare the number of errors before and after the current run. If an edit fixes some errors for a rule but introduces some errors for another rule, re-generating the baseline would cause the newly introduced errors to be ignored in subsequent runs. I'm not sure that makes sense.

Reply:
You are right. This assumes there is an additional control of the baseline, through PRs as an example.

There is a suggestion to change the approach to be more close to what `eslint-bulk` provides. In that case, a new option will be introduced called `--prune-suprressions` as briefly described [here](https://github.com/eslint/rfcs/pull/119#issuecomment-2263898275) which I believe covers this gap.